lighthouse-core/test/gather/gatherers/source-maps-test.js

```diff
@@ -0,0 +1,134 @@
+/**
+ * @license Copyright 2019 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+/* eslint-env jest */
+
+const SourceMaps = require('../../../gather/gatherers/source-maps.js');
+const path = require('path');
+const fs = require('fs');
+const pathToMap = path.join(__dirname,
+  '../../../../lighthouse-cli/test/fixtures/source-maps/bundle.js.map');
+const map = fs.readFileSync(pathToMap, 'utf-8');
+
+describe('SourceMaps gatherer', () => {
+  /**
+   * @param {{event: LH.Crdp.Debugger.ScriptParsedEvent, map: string, fetchError: string}} mapsAndEvents
+   * @return {Promise<LH.Artifacts['SourceMaps']>}
+   */
+  function getResults(mapsAndEvents) {
```

Comment:
I've been attempting to avoid bespoke mock/stub implementations in tests and use `jest.fn` wherever possible in new tests I write. I'm not sure if everyone else agrees but it'd be great to have comrades to make it a new standard going forward if you're interested 😃

Reply:
specifically can use `jest.fn().mockReturnValueOnce(firstVal).mockReturnValueOnce(secondVal).mockReturnValue(subsequentValues)` to model stuff pretty well.

Since it's just `evaluateAsync` and `sendCommand`, you may be better off mocking `Connection` that way and passing it to a real `Driver`, which will automatically handle creating events from protocol methods, registering the event listener, etc.

Call `SourceMaps.beforePass()`,`Debugger.enable` can be ignored, pass real events with the scripts back over the connection, call `SourceMaps.afterPass()`,`Debugger.enable` can be ignored, then pass back source maps  (or errors) from the `evaluateAsync` calls.

Reply:
I don't see how I can use a chain of `jest.fn().mockReturnValueOnce` to mock `Connection.sendCommand`, since there's some number of commands other than `Runtime.evaluate` to consider.

```
  /**
   * @param {{event: LH.Crdp.Debugger.ScriptParsedEvent, map: string, fetchError: string}} mapsAndEvents
   * @return {Promise<LH.Artifacts['SourceMaps']>}
   */
  function getResults(mapsAndEvents) {
    // Only the source maps that need to be fetched
    // (not 'sourceMappingURL=data:application/json;...') use the `evaluateAsync` code path.
    const mapsAndEventsForFetching =
      mapsAndEvents.filter(data => !data.event.sourceMapURL.startsWith('data:'));
    
    const connection = {
      evaluateAsync: jest.fn(),
      on: jest.fn(),
      sendCommand: jest.fn().mockReturnValue({}),
    };

    for (const mapAndEvent of mapsAndEventsForFetching) {
      if (mapAndEvent.map && mapAndEvent.fetchError) {
        throw new Error('should only define map or fetchError, not both.');
      }

      let result;
      if (mapAndEvent.fetchError) {
        result = {errorMessage: mapAndEvent.fetchError};
      } else {
        result = mapAndEvent.map;
      }
      connection.evaluateAsync = connection.evaluateAsync.mockReturnValueOnce({result});
    }

    const driver = new Driver(connection);

    const gatherer = new SourceMaps();
    gatherer._scriptParsedEvents = mapsAndEvents.map(data => data.event);
    return gatherer.afterPass({driver});
  }
```

![image](https://user-images.githubusercontent.com/4071474/60139382-85ddb700-9762-11e9-94cb-65584bc3d30a.png)

Reply:
i ended up stealing patrick's code :)